[SP][PCA] - Authenticate

Exceil-Matthew-Carpio opened the pull request:
- countryCode selection for currency BRL in response removed, and currency MYR added

rjaysantos reviewed (changes requested):
private function validateToken(Request $request, ?object $player): void
    {
        $playGame = $this->repository->getPlayGameByPlayIDToken(
            playID: $player->play_id,
            token: $request->externalToken
        );

        if (is_null($playGame) === true)
            throw new InvalidTokenException(request: $request);
    }

why nullable on $player parameter?

rjaysantos reviewed (changes requested):
public function authenticate(string $requestId, string $playID, string $currency): JsonResponse
    {
        $country = match ($currency) {
            'IDR' => 'ID',
            'PHP' => 'PH',
            'VND' => 'VN',
            'BRL' => 'BR',
            'USD' => 'US',
            'THB' => 'TH',
        };

        return response()->json(data: [
            "requestId" => $requestId,
            "username" => $playID,
            "currencyCode" => config('app.env') === 'PRODUCTION' ? $currency : 'CNY',
            "countryCode" => config('app.env') === 'PRODUCTION' ? $country : 'CN'
        ]);
    }

put the logic in the service

Exceil-Matthew-Carpio commented:
> private function validateToken(Request $request, ?object $player): void { $playGame = $this->repository->getPlayGameByPlayIDToken( playID: $player->play_id, token: $request->externalToken );
> 
> ```
>     if (is_null($playGame) === true)
>         throw new InvalidTokenException(request: $request);
> }
> ```
> 
> why nullable on $player parameter?

It should not be nullable. Process will stop from player validation when player is null, so it won't reach token validation. Will make adjustments accordingly

Exceil-Matthew-Carpio commented:
> public function authenticate(string $requestId, string $playID, string $currency): JsonResponse { $country = match ($currency) { 'IDR' => 'ID', 'PHP' => 'PH', 'VND' => 'VN', 'BRL' => 'BR', 'USD' => 'US', 'THB' => 'TH', };
> 
> ```
>     return response()->json(data: [
>         "requestId" => $requestId,
>         "username" => $playID,
>         "currencyCode" => config('app.env') === 'PRODUCTION' ? $currency : 'CNY',
>         "countryCode" => config('app.env') === 'PRODUCTION' ? $country : 'CN'
>     ]);
> }
> ```
> 
> put the logic in the service

Adjusted

Exceil-Matthew-Carpio commented:
> ```
>         "currencyCode" => config('app.env') === 'PRODUCTION' ? $countryData->currency : 'CNY',
>         "countryCode" => config('app.env') === 'PRODUCTION' ? $countryData->countryCode : 'CN'
> ```
> 
> including this. should be no logic regarding business on response class

Removed from response class and moved to service class

rjaysantos commented on Pca/PcaService.php:
```diff
@@ -105,13 +105,25 @@ private function getPlayerBalance(ICredentials $credentials, Request $request, s
         $this->validateToken(request: $request, player: $player);
 
-        return $player->currency;
+        $countryCode = match ($player->currency) {
+            'IDR' => 'ID',
+            'PHP' => 'PH',
+            'VND' => 'VN',
+            'USD' => 'US',
+            'THB' => 'TH',
+            'MYR' => 'MY'
+        };
+
+        return (object) [
+            'countryCode' => config('app.env') === 'PRODUCTION' ? $countryCode : 'CN',
+            'currency' => config('app.env') === 'PRODUCTION' ? $player->currency : 'CNY'
+        ];
```
just put it in credentials. then just return player credentials

rjaysantos commented on Pca/PcaController.php:
```diff
@@ -61,12 +61,13 @@ public function authenticate(Request $request)
             requestId: $request->requestId,
             playID: $request->username,
-            currency: $currency
+            currency: $playerCredentials->getCurrency(),
+            countryCode: $playerCredentials->getCountryCode()
```
just pass the credentials as a whole
